### PR TITLE
[AIRFLOW-3402] Support global k8s affinity and toleration configs

### PR DESCRIPTION
**Pull request by kppullin**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"

  - https://issues.apache.org/jira/browse/AIRFLOW-3402
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

- In the airflow.cfg file, adds new `affinity` and `tolerations` config keys to the `kubernetes` section.
  - The values for these are required to be in JSON format, as they are dense.
- Wires up support for `tolerations` for pods generated via the `KubernetesExecutor`.
- Minor updates to function documentation.

### Tests

- [X ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

  - Tests added to `test_kubernetes_executor.py` to validate the generated pod contains the defined affinity and tolerations config, for config coming from both the `kube_config` and `kube_executor_config` dicts.

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

- Documentation added to the config file and inline in tests.

### Code Quality

- [X] Passes `flake8`
- These changes at least don't make things worse...

### CODE REVIEW

**Comment by Fokko** [comment]
Sad CI:
```
======================================================================
40) FAIL: test_scheduler_sla_miss_callback_exception (tests.test_jobs.SchedulerJobTest)
----------------------------------------------------------------------
   Traceback (most recent call last):
    tests/test_jobs.py line 2852 in test_scheduler_sla_miss_callback_exception
      sla_callback.assert_called()
    .tox/py27-backend_mysql-env_docker/lib/python2.7/site-packages/mock/mock.py line 906 in assert_called
      raise AssertionError(msg)
   AssertionError: Expected 'None' to have been called.
   -------------------- >> begin captured stdout << ---------------------
   [2018-11-28 00:29:55,667] {test_task_view_type_check.py:47} INFO - class_instance type: <class 'unusual_prefix_61c0ab525b75d060ea41c0aa11a98c88efc72c26_test_task_view_type_check.CallableClass'>
   
   --------------------- >> end captured stdout << ----------------------
   -------------------- >> begin captured logging << --------------------
   root: INFO: class_instance type: <class 'unusual_prefix_61c0ab525b75d060ea41c0aa11a98c88efc72c26_test_task_view_type_check.CallableClass'>
   --------------------- >> end captured logging << ---------------------
======================================================================
41) FAIL: test_scheduler_sla_miss_email_exception (tests.test_jobs.SchedulerJobTest)
----------------------------------------------------------------------
   Traceback (most recent call last):
    .tox/py27-backend_mysql-env_docker/lib/python2.7/site-packages/mock/mock.py line 1305 in patched
      return func(*args, **keywargs)
    tests/test_jobs.py line 2896 in test_scheduler_sla_miss_email_exception
      'test_sla_miss')
    .tox/py27-backend_mysql-env_docker/lib/python2.7/site-packages/mock/mock.py line 925 in assert_called_with
      raise AssertionError('Expected call: %s\nNot called' % (expected,))
   AssertionError: Expected call: exception(u'Could not send SLA Miss email notification for DAG %s', u'test_sla_miss')
   Not called
```

**Comment by kppullin** [comment]
@Fokko - Thanks for the heads up on the CI failures!  I believe the referenced `sla_miss` failures were a red herring, with the actual errors being flake8 issues (my fault) and an existing python2 compatibility issue exposed by the new tests (lack of the `dict.copy` function).  I then broke things more with my first attempt at fixing the python2 issue, since I thought the affected data structure was a list.  CI went green once I switched the logic to handle a dict.

**Comment by kppullin** [comment]
@kaxil - Any chance to pick this up for 1.10.2?  It's the last piece needed on our end to isolate our airflow deployment within our kubernetes cluster.

File: airflow/contrib/kubernetes/pod.py [hunk 1]
```diff
@@ -60,6 +60,10 @@ class Pod:
     :type image_pull_secrets: str
     :param affinity: A dict containing a group of affinity scheduling rules
     :type affinity: dict
+    :param hostnetwork: If True enable host networking on the pod
+    :type hostnetwork: bool
+    :param tolerations: A list of kubernetes tolerations
+    :type tolerations: list tolerations
```
Comment (kaxil) [inline airflow/contrib/kubernetes/pod.py]: probably a typo. 
```suggestion
    :type tolerations: list
```

**Comment by kaxil** [comment]
@dimberman  @Fokko Any comments on this PR, please?

**Comment by dimberman** [comment]
@kaxil this LGTM but can we get this one merged first https://github.com/apache/incubator-airflow/pull/3770? It's good to go and I'd rather rebase this PR than that one (since the other one is a lot larger/been around longer).

**Comment by dimberman** [comment]
@kppullin looks like it's timing out on this test

```
test_pod_node_selectors (tests.contrib.minikube.test_kubernetes_pod_operator.KubernetesPodOperatorTest) ... 
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```


Seems like it could be related?

**Comment by kppullin** [comment]
@dimberman I've triggered a re-run of the tests, from the latest commit here (`e55f8c7`), on my personal repo and the test passed => https://travis-ci.com/kppullin/incubator-airflow/jobs/167696711

I don't know whether to chalk this up to an actual issue or a fluke with the k8s tests.

**Comment by Fokko** [comment]
I've restarted the test, let's see.

**Comment by kppullin** [comment]
Thanks @Fokko!  Looks like the rerun succeeded here as well.

**Review by Fokko — APPROVED** [review]
Thanks @kppullin

**Comment by kppullin** [comment]
@kaxil - Would this now be OK to cherry-pick into 1.10.2?

**Comment by kaxil** [comment]
Done

**Comment by kaxil** [comment]
Unfortunately, there are lots of conflicts & transitive dependencies when I cherry-pick it to v1.10-test branch, so I had to revert it from 1.10.2

Would appreciate any help to resolve it so that we can include it in 1.10.2 or else it would be available in next major release 1.11 or 2.0

**Pull request merged**